### PR TITLE
Refactored my Trie

### PR DESCRIPTION
**Pull request by Katzy**
Not sure why this was so challenging.  Conceptually, understanding what must happen was quite simple.  Looking back, when dealing with a Trie built upon a Tree built upon a LinkedList, knowing which type of node was going to be returned and how to manipulate the data properly with methods commensurate with the class proved to be a little bit of a headache.  Glad that's over.  On to the next.

### CODE REVIEW

File: minmax_stack.rb [hunk 1]
```diff
@@ -4,37 +4,60 @@
...
+
 class MinMaxStack
+
+  class DimensionError < StandardError;end
+
+  attr_accessor :items
+
   def initialize
+    @list = LinkedList.new
+    @items = []
```
Comment (jfarmer) [inline minmax_stack.rb]: There's no reason to use an `Array` here — you could just use another `LinkedList` with `unshift` instead of `push`.  Or better yet, a second `Stack`.

File: minmax_stack.rb [hunk 2]
```diff
@@ -4,37 +4,60 @@
...
+    self
   end
 
   # Removes the item on the top of the stack and returns it.
   # Raises an error if the stack is empty (called a "stack underflow")
   def pop
+    if self.empty?
+      raise DimensionError, "The stack is empty!"
+    else
+      @items = @items - [@list.head.value]
```
Comment (jfarmer) [inline minmax_stack.rb]: Oh.  Well, can't use a `LinkedList` if you're going to do this, obviously! :smile:

Comment (jfarmer) [inline minmax_stack.rb]: This isn't a great solution because this operation ("subtracting arrays") is likely `O(n)` in time, so you've now made the `#pop` method `O(n)` in time, too.

File: tree.rb [hunk 3]
```diff
@@ -25,5 +27,14 @@ def add_child(value)
   end
 
   def each(&block)
+    node = self.children
+    until node.empty?
+      block.call(node.head.value)
+      node = node.children
+    end
```
Comment (jfarmer) [inline tree.rb]: Are you sure this works?  Have you tested it?

File: trie.rb [hunk 4]
```diff
@@ -1,10 +1,52 @@
...
+      temp_node = find(item).value
+      item = item.sub(self.prefix, "")
+    end
+    until item.length == 0
+      temp_node.add_child(item[0])
+      prefix <<  item[0]
+      item = item[1..-1]
+      temp_node = temp_node.children.head.value
+    end
+    p prefix
```
Comment (jfarmer) [inline trie.rb]: remove debugging code

File: trie.rb [hunk 5]
```diff
@@ -1,10 +1,52 @@
 require_relative "tree"
 
-# Implement a Trie (sometimes called a "prefix tree")
-# See: http://en.wikipedia.org/wiki/Trie
-#
-# A Trie allows for efficient prefix searching, i.e.,
-# "Find me all things that start with 'Foo'"
-
 class Trie
+  attr_reader :root, :prefix
```
Comment (jfarmer) [inline trie.rb]: Does this work?  The code is kind of hard to follow and there are parts of it that are suspect to me.

File: tree.rb (hunk 3 above)
Comment (Katzy) [inline tree.rb]: Actually...  this code existed already.  I didn't write this code.  so... no.. i don't think it works.  :)  actually, I'm glad you asked that.  it's why I didn't use Tree.each in my Trie.    anyway...   hi

Comment (jfarmer) [inline tree.rb]: `self.children` returns a list of nodes and it seems like you're only ever passing the first child to `block.call`.

File: tree.rb [hunk 6]
```diff
@@ -1,5 +1,7 @@
 require_relative "linked_list"
 
+require 'pp'
```
Comment (jfarmer) [inline tree.rb:3]: remove debugging code

File: minmax_stack.rb [hunk 7]
```diff
@@ -4,37 +4,60 @@
...
 
+
 class MinMaxStack
+
+  class DimensionError < StandardError;end
+
+  attr_accessor :items
+
   def initialize
+    @list = LinkedList.new
```
Comment (jfarmer) [inline minmax_stack.rb]: There's no reason to use a `LinkedList` here — you should use a `Stack` (which you've already implemented).

File: minmax_stack.rb [hunk 8]
```diff
@@ -4,37 +4,60 @@
...
   end
 
   # Return the number of items on the stack
   def size
+    @list.length
   end
 
   # Returns the smallest item on the stack
   def min
+    @items.min
```
Comment (jfarmer) [inline minmax_stack.rb]: This is not constant time.

File: minmax_stack.rb [hunk 9]
```diff
@@ -4,37 +4,60 @@
...
   end
 
   # Returns the largest item on the stack
   # O(1) time
   def max
+    @items.max
```
Comment (jfarmer) [inline minmax_stack.rb]: This is not constant time.

File: minmax_stack.rb [hunk 10]
```diff
@@ -4,37 +4,60 @@
...
+
   def initialize
+    @list = LinkedList.new
+    @items = []
   end
 
   # Places +item+ on the top of the stack
   def push(item)
+    @list.unshift(item)
+    @items.push(item)
```
Comment (jfarmer) [inline minmax_stack.rb]: It's possible to implement a MinMaxStack using three separate stacks, but first nail the MinStack and MaxStack implementations.  It'll be clear what needs to happen in MinMaxStack when the other two are solid.  Are those in your code at all?  I don't see them in this PR.
